js/views/components/CryptoTradingPair.js

```diff
@@ -0,0 +1,125 @@
+import { ensureMainnetCode } from '../../utils/crypto';
+import loadTemplate from '../../utils/loadTemplate';
+import BaseVw from '../baseVw';
+
+export default class extends BaseVw {
+  constructor(options = {}) {
+    const serverCur = getServerCurrency();
+    const opts = {
+      ...options,
+      initialState: {
+        tradingPairClass: 'cryptoTradingPairLg',
+        exchangeRateClass: '',
+        fromCur: serverCur && serverCur.code || '',
+        fromCurAmount: 1,
+        toCur: '',
+        localCurrency: app.settings.get('localCurrency'),
+        // If passing this in, it should be a string or a function. If it's a function
+        // it will be passed the coinType and should return a string.
+        noExchangeRateTip: coinType =>
+          app.polyglot.t('cryptoTradingPair.tipMissingExchangeRate',
+              { coinType }),
+        exchangeRateUnavailable: false,
+        iconClass: 'ion-alert-circled clrTAlert',
+        ...options.initialState || {},
+      },
+    };
+
+    super(opts);
+
+    // Since the initial state is not being piped through setState in the
+    // base class, this is a hack to run it through setState now and ensure
+    // setState updates the exchange rate which is based on the toCur changing.
+    if (typeof this._state.toCur === 'string' && this._state.toCur) {
+      const toCur = this._state.toCur;
+      this._state.toCur = '';
+      this.setState({
+        ...this._state,
+        toCur,
+      });
+    }
+
+    this.listenTo(currencyEvents, 'exchange-rate-change', e => {
+      if (
+        e.changed.includes(this.getState().toCur) ||
+        e.changed.includes(this.getState().fromCur)
+      ) {
+        const curState = this.getState();
+        this.setState({
+          ...curState,
+          ...this.getConversionState(curState.fromCur, curState.toCur, curState.fromCurAmount),
+        });
+      }
+    });
+  }
+
+  className() {
+    return 'cryptoTradingPairWrap';
+  }
+
+  setState(state = {}, options = {}) {
+    const prevState = this.getState();
+    let newState = {
+      ...prevState,
+      ...state,
+    };
+
+    if (typeof state === 'object') {
+      if (typeof state.fromCur === 'string') {
+        newState.fromCur = ensureMainnetCode(state.fromCur.toUpperCase());
+      }
+
+      if (typeof state.toCur === 'string') {
+        newState.toCur = ensureMainnetCode(state.toCur.toUpperCase());
+      }
+
+      if (newState.fromCur !== prevState.fromCur ||
+        newState.toCur !== prevState.toCur ||
+        newState.fromCurAmount !== prevState.fromCurAmount) {
+        newState = {
+          ...newState,
+          ...this.getConversionState(newState.fromCur, newState.toCur, newState.fromCurAmount),
+        };
+      }
+    }
+
+    return super.setState(newState, options);
+  }
+
+  getConversionState(fromCur, toCur, fromCurAmount) {
+    const exchangeRate = getExchangeRate(toCur);
+    const toCurAmount = fromCurAmount * getExchangeRate(toCur);
```

Comment:
> Can you multiply by exchangeRate here, since it's getExchangeRate(toCur)?

Yes, i can. good catch.

> Is NaN an ok value for it?

I think it is. If there's any confusion as to why it would be NaN, the exchangeRateUnavailable being true would IMO make it clear as to why. This isn't really a public method either, it's just used to derive the state of this view and I would guess the template follows a different code path if exchangeRateUnavailable is true and doesn't even bother with the fromCurConvertedAmount variable in that case.